flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeRowsUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,719 @@
+import org.apache.flink.table.runtime.generated.RecordEqualiser;
+import org.apache.flink.table.runtime.keyselector.RowDataKeySelector;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.table.types.logical.BigIntType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.types.RowKind;
+import org.apache.flink.util.Collector;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+import static org.apache.flink.table.runtime.util.StateConfigUtil.createTtlConfig;
+
+/**
+ * The NonTimeRowsUnboundedPrecedingFunction class is a specialized implementation for processing
+ * unbounded OVER window aggregations, particularly for non-time-based rows queries in Apache Flink.
+ * It maintains strict ordering of rows within partitions and handles the full changelog lifecycle
+ * (inserts, updates, deletes).
+ *
+ * <p>Key Components and Assumptions
+ *
+ * <p>Data Structure Design: (1) Maintains a sorted list of tuples containing sort keys and lists of
+ * IDs for each key (2) Each incoming row is assigned a unique Long ID (starting from
+ * Long.MIN_VALUE) (3) Uses multiple state types to track rows, sort orders, and aggregations
+ *
+ * <p>State Management: (1) idState: Counter for generating unique row IDs (2) sortedListState:
+ * Ordered list of sort keys with their associated row IDs (3) valueMapState: Maps IDs to their
+ * corresponding input rows (4) accMapState: Maps sort keys to their accumulated values
+ *
+ * <p>Processing Model: (1) For inserts/updates: Adds rows to the appropriate position based on sort
+ * key (2) For deletes: Removes rows by matching both sort key and row content (3) Recalculates
+ * aggregates for affected rows and emits the appropriate events (4) Skips redundant events when
+ * accumulators haven't changed to reduce network traffic
+ *
+ * <p>Optimization Assumptions: (1) Skip emitting updates when accumulators haven't changed to
+ * reduce network traffic (2) Uses state TTL for automatic cleanup of stale data (3) Carefully
+ * manages row state to support incremental calculations
+ *
+ * <p>Retraction Handling: (1) Handles retraction mode (DELETE/UPDATE_BEFORE) events properly (2)
+ * Supports the proper processing of changelog streams
+ *
+ * <p>Limitations
+ *
+ * <p>Linear search performance: - The current implementation uses a linear search to find the
+ * correct position for each sort key. This can be optimized using a binary search for large state
+ * sizes.
+ *
+ * <p>State size and performance: - The implementation maintains multiple state types that could
+ * grow large with high cardinality data
+ *
+ * <p>Linear recalculation: - When processing updates, all subsequent elements need to be
+ * recalculated, which could be inefficient for large windows
+ */
+public class NonTimeRowsUnboundedPrecedingFunction<K>
```

Comment:
this new class appears to have a lot of duplicate code to the Range class. I suggest that the code be organised so that there is minimal duplicate code. Maybe a super class ?

Reply:
@davidradl Thanks for taking the time to review. I accidentally created a PR instead of draft PR. I have a separate commit which refactors Range vs Rows but Im still working on it.

Reply:
Introduced `AbstractNonTimeUnboundedPrecedingOver`